### PR TITLE
Refactor "maps" of annotations/threads within the `selection` store module

### PR DESCRIPTION
**Pull request by lyzadanger**
There are several properties in the `selection` store state that are kind of map-like, kind of Set-like. These include the currently-selected annotations, the currently-highlighted annotations, a map of expanded/collapsed threads, etc. There has been some inconsistency both in their internal representation and external representation (via selectors).

This PR aims to make consistent the maplike collections in the `selection` state such that they are predictable and conventionalized. For each of the affected collections, the internal structure is represented as `{Object<string,boolean>}` where the string key is an identifier (id or tag) of an annotation or thread. In several of these maps, entries are always `true`, making them feel like Arrays, but for now I've opted to make them all Objects, because:

* As in a Set, you'll end up with only one entry per annotation ID (protects from dupes)
* It may be useful, now or ever, to be able to set an entry to `false` — this is already useful for the `expanded` collection, e.g. Even if not, it's nice to have these all consistent.

The selector conventions are:

* `<nameOfProperty>Annotations`, e.g. `forcedVisibleAnnotations`, `selectedAnnotations`: Returns an `Array` of all annotation identifiers in the map object with a `true` value.
* `<nameOfProperty>Map`, e.g. `expandedMap`: Retrieves the map object itself, with all entries

And for actions:

* `set<nameOfProperty>`, e.g. `setExpanded`, `setForcedVisible`. takes single `id` and boolean: _updates_ current map
* `<verbFormOfProperty>Annotations`, e.g. `highlightAnnotations`, `selectAnnotations`. _replaces_ current map. Takes Array of `id`s and sets them all to `true` in the new map.

I've only added actions and selectors for these where they're actively needed. Figure we can add them just in time as we go.

More work to do in this module, but this is plenty chunky for a PR.

### CODE REVIEW

File: src/sidebar/build-thread.js [hunk 1]
```diff
@@ -263,7 +263,7 @@ function hasVisibleChildren(thread) {
  * @typedef Options
  * @prop {string[]} [selected] - List of currently-selected annotation ids, from
  *       the data store
- * @prop {string[]} [forceVisible] - List of ids of annotations that have
+ * @prop {string[]} [forcedVisible] - List of ids of annotations that have
```
Comment (lyzadanger) [inline src/sidebar/build-thread.js:266]: `forcedVisible` is a better adjectival form than `forceVisible`. `forceVisible` is the verb form (and is appropriate for function naming).

File: src/sidebar/components/annotation-header.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ export default function AnnotationHeader({
   threadIsCollapsed,
 }) {
   const isCollapsedReply = isReply(annotation) && threadIsCollapsed;
-  const setCollapsed = useStore(store => store.setCollapsed);
+  const setExpanded = useStore(store => store.setExpanded);
```
Comment (lyzadanger) [inline src/sidebar/components/annotation-header.js:28]: Here and in several other components/consumers, `setExpanded` makes more semantic sense than `setCollapsed` because almost all users are trying to expand, not trying to collapse. It also helps to delineate a little from the `thread.collapsed` property, which is informed by, but not entirely dictated by, the values in the resulting map.

File: src/sidebar/components/search-status-bar.js [hunk 3]
```diff
@@ -41,13 +41,13 @@ function SearchStatusBar({ rootThread }) {
...
     focusModeUserPrettyName,
-    selectionMap,
+    hasSelectedAnnotations,
     selectedTab,
   } = useStore(store => ({
     filterQuery: store.getState().selection.filterQuery,
     focusModeFocused: store.focusModeFocused(),
     focusModeUserPrettyName: store.focusModeUserPrettyName(),
-    selectionMap: store.getSelectedAnnotationMap(),
+    hasSelectedAnnotations: store.hasSelectedAnnotations(),
```
Comment (lyzadanger) [inline src/sidebar/components/search-status-bar.js:50]: Use selector instead of grabbing state directly here...plus, what this consumer actually cares about is whether there are any selected annotations, not what they are.

File: src/sidebar/services/root-thread.js [hunk 4]
```diff
@@ -56,10 +50,10 @@ export default function RootThread(
     };
 
     const options = {
-      forceVisible: truthyKeys(state.selection.forceVisible),
-      expanded: store.expandedThreads(),
+      forcedVisible: store.forcedVisibleAnnotations(),
```
Comment (lyzadanger) [inline src/sidebar/services/root-thread.js:53]: Don't grab state directly here.

File: src/sidebar/services/root-thread.js [hunk 5]
```diff
@@ -56,10 +50,10 @@ export default function RootThread(
...
+      expanded: store.expandedMap(),
       highlighted: state.selection.highlighted,
-      selected: truthyKeys(store.getSelectedAnnotationMap() || {}),
+      selected: store.selectedAnnotations(),
```
Comment (lyzadanger) [inline src/sidebar/services/root-thread.js:56]: Don't grab state; use a selector.

File: src/sidebar/services/root-thread.js [hunk 6]
```diff
@@ -56,10 +50,10 @@ export default function RootThread(
     };
 
     const options = {
-      forceVisible: truthyKeys(state.selection.forceVisible),
-      expanded: store.expandedThreads(),
+      forcedVisible: store.forcedVisibleAnnotations(),
+      expanded: store.expandedMap(),
       highlighted: state.selection.highlighted,
```
Comment (lyzadanger) [inline src/sidebar/services/root-thread.js:55]: Crap, overlooked this one.

File: src/sidebar/services/test/root-thread-test.js [hunk 7]
```diff
@@ -56,13 +55,13 @@ describe('rootThread', function () {
         return this.state;
       },
 
-      expandedThreads: sinon.stub().returns({}),
-      getSelectedAnnotationMap: sinon.stub().returns(null),
+      expandedMap: sinon.stub().returns({}),
```
Comment (lyzadanger) [inline src/sidebar/services/test/root-thread-test.js:58]: Same result, just renamed.

File: src/sidebar/store/modules/selection.js [hunk 8]
```diff
@@ -18,8 +13,7 @@ import { createSelector } from 'reselect';
 
 import uiConstants from '../../ui-constants';
 import * as metadata from '../../util/annotation-metadata';
-import { countIf, toSet } from '../../util/array';
-import immutable from '../../util/immutable';
+import { countIf } from '../../util/array';
```
Comment (lyzadanger) [inline src/sidebar/store/modules/selection.js:16]: Was able to remove `immutable`, `freeze` legacy stuff.

File: src/sidebar/store/modules/selection.js [hunk 9]
```diff
@@ -51,20 +45,26 @@ TAB_SORTKEYS_AVAILABLE[uiConstants.TAB_ORPHANS] = [
...
 ];
 
+/**
+ * Utility function that returns all of the properties of an object whose
+ * value is `true`.
+ *
+ * @param {Object} obj
+ * @return {any[]}
+ */
+function truthyKeys(obj) {
```
Comment (lyzadanger) [inline src/sidebar/store/modules/selection.js:55]: I realized this name is a little silly: it doesn't return for "truthy" entries, it requires literal `true`. I copied it from elsewhere...

File: src/sidebar/store/modules/selection.js [hunk 10]
```diff
@@ -51,20 +45,26 @@ TAB_SORTKEYS_AVAILABLE[uiConstants.TAB_ORPHANS] = [
...
+  return Object.keys(obj).filter(key => obj[key] === true);
+}
+
 function initialSelection(settings) {
   const selection = {};
+  // TODO: Do not take into account existence of `settings.query` here
```
Comment (lyzadanger) [inline src/sidebar/store/modules/selection.js:61]: Something to take care of in a later iteration...

File: src/sidebar/store/modules/selection.js [hunk 11]
```diff
@@ -51,20 +45,26 @@ TAB_SORTKEYS_AVAILABLE[uiConstants.TAB_ORPHANS] = [
...
-  return freeze(selection);
-}
-
-function freeze(selection) {
-  if (Object.keys(selection).length) {
-    return immutable(selection);
-  } else {
-    return null;
-  }
+  return selection;
```
Comment (lyzadanger) [inline src/sidebar/store/modules/selection.js:67]: Never `null` anymore.

File: src/sidebar/store/modules/selection.js [hunk 12]
```diff
@@ -395,26 +419,35 @@ function setSortKey(key) {
...
 
-/** Is the annotation referenced by `$tag` currently focused? */
-function isAnnotationFocused(state, $tag) {
-  return state.selection.focusedAnnotations.some(
-    focusedAnn => $tag && focusedAnn === $tag
-  );
+/* Selectors */
+
+function focusedAnnotations(state) {
+  return truthyKeys(state.selection.focused);
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:425]: Depending on how often these selectors are invoked per render, you might want to use `createSelector` to memoize the result. eg.

```
const focusedAnnotations = createSelector(
  state => state.selection.focused,
  focused => truthyKeys(focused))
);
```

File: src/sidebar/store/modules/selection.js [hunk 13]
```diff
@@ -51,20 +45,26 @@ TAB_SORTKEYS_AVAILABLE[uiConstants.TAB_ORPHANS] = [
   'Location',
 ];
 
+/**
+ * Utility function that returns all of the properties of an object whose
+ * value is `true`.
+ *
+ * @param {Object} obj
+ * @return {any[]}
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:53]:
```suggestion
 * @return {string[]}
```

All keys are strings, so you can be more specific here.

File: src/sidebar/store/modules/selection.js [hunk 14]
```diff
@@ -92,29 +92,40 @@ const setTab = (selectedTab, newTab) => {
...
+    /**
+     * The following objects map annotation identifiers to a boolean
+     * (typically `true`). They are objects (i.e. instead of Arrays) for two
+     * reasons:
+     * - Allows explicit setting of `false`
+     * - Prevents duplicate entries for a single annotation
+     */
+
+    // A set of annotations that are currently "focused" — hovered over in
+    // the UI, e.g.
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:104]: Was there a comment missing after `e.g.`?

File: src/sidebar/store/modules/selection.js [hunk 15]
```diff
@@ -92,29 +92,40 @@ const setTab = (selectedTab, newTab) => {
...
+    // A set of annotations that are currently "selected" by the user —
+    // these will supersede other filters/selections
+    selected: initialSelection(settings),
+
+    // Explicitly-expanded or -collapsed annotations (threads). A collapsed
+    // annotation thread will not show its replies; an expanded thread will
+    // show its replies. Note that there are other factors affecting
+    // collapsed states, e.g., top-level threads are collapsed by default
+    // until explicitly expanded.
     expanded: initialSelection(settings) || {},
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:116]: `initialSelection()` always returns an object, so the `|| {}` is redundant.

File: src/sidebar/store/modules/selection.js [hunk 16]
```diff
@@ -138,22 +149,32 @@ const update = {
...
+    action.toggleIds.forEach(id => {
+      selection[id] = !selection[id];
+    });
+    return { selected: selection };
   },
 
   FOCUS_ANNOTATIONS: function (state, action) {
-    return { focusedAnnotations: [...action.focusedTags] };
+    const focused = {};
+    action.focusedTags.forEach(tag => (focused[tag] = true));
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:176]: The logic of turning a list of strings into an `Object.<string,true>` set is repeated a few times. This could perhaps be moved to a helper.

File: src/sidebar/store/modules/selection.js [hunk 17]
```diff
@@ -257,63 +280,57 @@ const update = {
...
-    for (let i = 0; i < ids.length; i++) {
-      const id = ids[i];
-      if (selection[id]) {
-        delete selection[id];
-      } else {
-        selection[id] = true;
-      }
-    }
-    dispatch(select(selection));
+function toggleSelectedAnnotations(toggleIds) {
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:304]: Eek! - The previous code here was pretty complex for what it did.

File: src/sidebar/store/modules/test/selection-test.js [hunk 18]
```diff
@@ -28,37 +28,51 @@ describe('sidebar/store/modules/selection', () => {
     });
   });
 
-  describe('setForceVisible()', function () {
+  describe('setForcedVisible()', function () {
```
Comment (robertknight) [inline src/sidebar/store/modules/test/selection-test.js:31]:
```suggestion
  describe('setForcedVisible', function () {
```

describe blocks shouldn't include parens, just the function name.

File: src/sidebar/components/thread.js [hunk 19]
```diff
@@ -40,7 +40,7 @@ function Thread({ showDocumentInfo = false, thread, threadsService }) {
     child => countVisible(child) > 0
   );
 
-  const onToggleReplies = () => setCollapsed(thread.id, !thread.collapsed);
+  const onToggleReplies = () => setExpanded(thread.id, !!thread.collapsed);
```
Comment (robertknight) [inline src/sidebar/components/thread.js:43]:
```suggestion
  const onToggleReplies = () => setExpanded(thread.id, !!thread.collapsed);
```

Is the `!!` needed here? If we can be sure that `thread.collapsed` is actually a boolean then it shouldn't be.

File: src/sidebar/store/modules/selection.js [hunk 20]
```diff
@@ -257,63 +280,57 @@ const update = {
 
 const actions = util.actionTypes(update);
 
-function select(annotations) {
+/**
+ * Set the currently selected annotation IDs. This will replace the current
+ * selection. All provided annotation ids will be set to `true` in the selection.
+ *
+ * @param {string[]} ids - Idenfiers of annotations to select
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:287]: Man, I made this typo like six times while working on this set of changes...what's _with_ that?

File: src/sidebar/store/modules/selection.js (hunk 14 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:104]: No, but my grammar may be a little hard to parse; I'll rewrite as `e.g. hovered over in the UI`

File: src/sidebar/store/modules/selection.js (hunk 17 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:304]: I know, rite?!

File: src/sidebar/store/modules/selection.js (hunk 12 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:425]: And done...

File: src/sidebar/store/modules/selection.js (hunk 16 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:176]: Thanks, did it.
